stats/csv/collector_test.go

```diff
@@ -0,0 +1,328 @@
+func TestRunCollect(t *testing.T) {
+	testSamples := []stats.SampleContainer{
+		stats.Sample{
+			Time:   time.Unix(1562324643, 0),
+			Metric: stats.New("my_metric", stats.Gauge),
+			Value:  1,
+			Tags: stats.NewSampleTags(map[string]string{
+				"tag1": "val1",
+				"tag2": "val2",
+				"tag3": "val3",
+			}),
+		},
+		stats.Sample{
+			Time:   time.Unix(1562324644, 0),
+			Metric: stats.New("my_metric", stats.Gauge),
+			Value:  1,
+			Tags: stats.NewSampleTags(map[string]string{
+				"tag1": "val1",
+				"tag2": "val2",
+				"tag3": "val3",
+				"tag4": "val4",
+			}),
+		},
+	}
+
+	t.Run("Run and Collect", func(t *testing.T) {
+		mem := afero.NewMemMapFs()
+		collector, err := New(mem, "path", lib.TagSet{"tag1": true, "tag2": false, "tag3": true})
+		assert.NoError(t, err)
+		assert.NotNil(t, collector)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			collector.Run(ctx)
+			wg.Done()
+		}()
+		err = collector.Init()
+		assert.NoError(t, err)
+		collector.Collect(testSamples)
+		time.Sleep(1 * time.Second)
+		cancel()
+		wg.Wait()
+		csvbytes, _ := afero.ReadFile(mem, "path")
+		csvstr := fmt.Sprintf("%s", csvbytes)
+		assert.Equal(t,
+			"metric_name,timestamp,metric_value,tag1,tag3,extra_tags\n"+
+				"my_metric,1562324643,1.000000,val1,val3,\n"+
+				"my_metric,1562324644,1.000000,val1,val3,tag4=val4\n",
+			csvstr)
+	})
+}
+
+func TestNew(t *testing.T) {
+	configs := []struct {
+		fname string
+		tags  lib.TagSet
+	}{
+		{
+			fname: "name",
+			tags: lib.TagSet{
+				"tag1": true,
+				"tag2": false,
+				"tag3": true,
+			},
+		},
+		{
+			fname: "-",
+			tags: lib.TagSet{
+				"tag1": true,
+			},
+		},
+		{
+			fname: "",
+			tags: lib.TagSet{
+				"tag1": false,
+				"tag2": false,
+			},
+		},
+	}
+	expected := []struct {
+		fname       string
+		resTags     []string
+		ignoredTags []string
+	}{
+		{
+			fname: "name",
+			resTags: []string{
+				"tag1", "tag3",
+			},
+			ignoredTags: []string{
+				"tag2",
+			},
+		},
+		{
+			fname: "-",
+			resTags: []string{
+				"tag1",
+			},
+			ignoredTags: []string{},
+		},
+		{
+			fname:   "-",
+			resTags: []string{},
+			ignoredTags: []string{
+				"tag1", "tag2",
+			},
+		},
+	}
+
+	for i := range configs {
+		func(config struct {
+			fname string
+			tags  lib.TagSet
+		}, expected struct {
+			fname       string
+			resTags     []string
+			ignoredTags []string
+		}) {
+			t.Run(config.fname, func(t *testing.T) {
+				collector, err := New(afero.NewMemMapFs(), config.fname, config.tags)
+				assert.NoError(t, err)
+				assert.NotNil(t, collector)
+				assert.Equal(t, expected.fname, collector.fname)
+				sort.Strings(expected.resTags)
+				sort.Strings(collector.resTags)
+				assert.Equal(t, expected.resTags, collector.resTags)
+				sort.Strings(expected.ignoredTags)
+				sort.Strings(collector.ignoredTags)
+				assert.Equal(t, expected.ignoredTags, collector.ignoredTags)
+			})
+		}(configs[i], expected[i])
```

Comment:
You could just do `config, expected := configs[i], expected[i]` instead of the function call. 
It will both be much shorter and much easier to support - if you want to add a new field you will now need to add it twice